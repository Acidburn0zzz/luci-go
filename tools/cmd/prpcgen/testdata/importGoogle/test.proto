// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto3";

package test;

import "google/protobuf/descriptor.proto";
import "google/protobuf/timestamp.proto";

service Greeter {
  rpc Describe (Void) returns (google.protobuf.FileDescriptorProto) {}
  rpc Timestamp (Void) returns (google.protobuf.Timestamp) {}
}

message Void {}
