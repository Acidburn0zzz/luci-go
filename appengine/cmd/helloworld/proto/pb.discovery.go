// AUTOGENERATED. DO NOT EDIT.

package helloworld

import discovery "github.com/luci/luci-go/server/discovery"

func init() {
	discovery.RegisterDescriptorSetCompressed(
		[]string{
			"helloworld.Greeter",
		},
		[]byte{31, 139,
			8, 0, 0, 9, 110, 136, 0, 255, 116, 147, 193, 110, 211, 78,
			16, 198, 189, 206, 214, 117, 38, 233, 63, 238, 254, 237, 170, 138,
			144, 24, 229, 64, 57, 32, 211, 6, 193, 5, 129, 84, 85, 21,
			20, 85, 28, 90, 202, 221, 141, 39, 137, 37, 215, 27, 118, 157,
			162, 190, 18, 55, 14, 60, 24, 111, 192, 236, 54, 134, 30, 224,
			230, 207, 251, 205, 124, 191, 157, 177, 225, 123, 4, 201, 146, 234,
			90, 127, 213, 166, 46, 243, 149, 209, 173, 86, 240, 231, 205, 100,
			2, 195, 247, 78, 93, 208, 151, 53, 217, 86, 41, 144, 77, 113,
			67, 251, 2, 197, 211, 254, 133, 127, 158, 60, 1, 216, 120, 86,
			245, 157, 218, 135, 237, 27, 178, 182, 88, 116, 166, 78, 78, 207,
			96, 251, 157, 33, 106, 201, 168, 183, 16, 95, 22, 119, 190, 74,
			237, 231, 15, 8, 30, 134, 141, 247, 254, 114, 194, 17, 147, 224,
			195, 79, 9, 145, 146, 50, 216, 19, 240, 67, 128, 24, 170, 158,
			12, 212, 244, 155, 192, 19, 189, 186, 51, 213, 98, 217, 226, 244,
			240, 232, 21, 126, 90, 18, 158, 95, 157, 156, 225, 241, 186, 93,
			106, 99, 115, 60, 174, 107, 244, 6, 139, 134, 44, 153, 91, 42,
			115, 192, 43, 75, 168, 231, 216, 46, 43, 139, 86, 175, 205, 140,
			112, 166, 75, 66, 150, 11, 125, 75, 166, 161, 18, 215, 77, 73,
			134, 45, 132, 199, 171, 98, 230, 26, 87, 51, 106, 44, 61, 195,
			207, 100, 108, 165, 27, 156, 230, 135, 192, 134, 162, 197, 89, 209,
			224, 53, 225, 92, 115, 17, 86, 141, 175, 58, 63, 59, 57, 253,
			120, 121, 138, 243, 170, 166, 28, 32, 6, 17, 170, 94, 20, 43,
			200, 33, 140, 2, 37, 251, 193, 80, 140, 39, 30, 121, 225, 198,
			84, 53, 11, 116, 128, 156, 130, 37, 205, 171, 166, 106, 57, 132,
			43, 1, 122, 81, 32, 84, 175, 31, 143, 224, 49, 200, 40, 8,
			3, 213, 27, 132, 47, 199, 10, 47, 169, 41, 45, 22, 191, 27,
			0, 12, 97, 203, 25, 216, 62, 136, 254, 235, 20, 7, 15, 146,
			71, 157, 234, 177, 58, 56, 130, 55, 16, 242, 16, 229, 40, 216,
			21, 227, 35, 143, 97, 238, 247, 128, 155, 5, 242, 72, 154, 182,
			96, 14, 6, 115, 23, 90, 51, 221, 129, 69, 247, 13, 220, 83,
			73, 23, 51, 138, 83, 24, 0, 239, 198, 81, 37, 161, 130, 29,
			216, 114, 66, 42, 153, 132, 163, 61, 23, 234, 228, 150, 59, 140,
			59, 197, 117, 73, 127, 167, 83, 12, 148, 36, 187, 240, 154, 129,
			132, 146, 41, 111, 121, 252, 124, 3, 100, 87, 154, 71, 254, 47,
			162, 238, 218, 246, 30, 71, 112, 219, 148, 7, 236, 112, 132, 195,
			201, 194, 204, 227, 8, 143, 147, 133, 105, 234, 35, 133, 199, 201,
			54, 56, 194, 227, 100, 253, 164, 83, 140, 147, 253, 159, 94, 71,
			254, 239, 120, 241, 43, 0, 0, 255, 255, 98, 72, 185, 105, 52,
			3, 0, 0},
	)
}
