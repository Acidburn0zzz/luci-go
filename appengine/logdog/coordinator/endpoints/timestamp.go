// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package endpoints

import (
	"time"
)

// ToRFC3339 converts a time.Time to an RFC3339 timestamp string.
func ToRFC3339(t time.Time) string {
	return t.Format(time.RFC3339Nano)
}

// ParseRFC3339 parses an RFC3339 string into a time.Time.
func ParseRFC3339(value string) (time.Time, error) {
	if value == "" {
		return time.Time{}, nil
	}
	t, err := time.Parse(time.RFC3339Nano, value)
	if err != nil {
		return time.Time{}, err
	}
	return t, nil
}
