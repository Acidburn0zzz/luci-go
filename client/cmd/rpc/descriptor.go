// Copyright 2016 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package main

import (
	"fmt"
	"strings"

	"github.com/golang/protobuf/proto"
	"golang.org/x/net/context"

	"github.com/luci/luci-go/common/proto/google/descriptor"
	"github.com/luci/luci-go/common/prpc"
	"github.com/luci/luci-go/server/discovery"
)

type serverDescription struct {
	services   []string
	descriptor *descriptor.FileDescriptorSet
}

func loadDescription(c context.Context, client *prpc.Client) (*serverDescription, error) {
	dc := discovery.NewDiscoveryPRPCClient(client)
	res, err := dc.Describe(c, &discovery.Void{})
	if err != nil {
		return nil, fmt.Errorf("could not load server description: %s", err)
	}

	result := &serverDescription{
		services:   res.Services,
		descriptor: &descriptor.FileDescriptorSet{},
	}
	if err := proto.Unmarshal(res.FileDescriptionSet, result.descriptor); err != nil {
		return nil, fmt.Errorf("could not unmarshal FileDescriptionSet: %s", err)
	}
	return result, nil
}

// resolveInputMessage resolves input message type of a method.
func (d *serverDescription) resolveInputMessage(service, method string) (*descriptor.DescriptorProto, error) {
	_, obj, _ := d.descriptor.Resolve(service)
	serviceDesc, ok := obj.(*descriptor.ServiceDescriptorProto)
	if !ok {
		return nil, fmt.Errorf("service %q not found", service)
	}

	mi := serviceDesc.FindMethod(method)
	if mi == -1 {
		return nil, fmt.Errorf("method %q in service %q not found", method, service)
	}

	msgName := serviceDesc.Method[mi].GetInputType()
	msgName = strings.TrimPrefix(msgName, ".")
	return d.resolveMessage(msgName)
}

func (d *serverDescription) resolveMessage(name string) (*descriptor.DescriptorProto, error) {
	_, obj, _ := d.descriptor.Resolve(name)
	msg, ok := obj.(*descriptor.DescriptorProto)
	if !ok {
		return nil, fmt.Errorf("message %q not found", name)
	}
	return msg, nil
}
