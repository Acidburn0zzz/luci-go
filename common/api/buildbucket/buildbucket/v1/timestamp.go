// Copyright 2018 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package buildbucket

import "time"

// ParseTimestamp parses a buildbucket timestamp.
func ParseTimestamp(usec int64) time.Time {
	if usec == 0 {
		return time.Time{}
	}
	// TODO(nodir): will start to overflow in year ~2250+, fix it then.
	return time.Unix(0, usec*1000).UTC()
}

// FormatTimestamp converts t to a buildbucket timestamp.
func FormatTimestamp(t time.Time) int64 {
	return t.UnixNano() / 1000
}
