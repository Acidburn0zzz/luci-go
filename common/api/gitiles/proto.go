// Copyright 2017 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package gitiles

import (
	"encoding/hex"
	"time"

	"github.com/golang/protobuf/ptypes"

	"go.chromium.org/luci/common/errors"
	"go.chromium.org/luci/common/proto/git"
)

// Proto converts this User to its protobuf equivalent.
func (u *User) Proto() (ret *git.Commit_User, err error) {
	ret = &git.Commit_User{
		Name:  u.Name,
		Email: u.Email,
	}
	if u.Time != "" {
		var ts time.Time
		if ts, err = u.GetTime(); err != nil {
			err = errors.Annotate(err, "decoding time").Err()
			return
		}
		ret.Time, err = ptypes.TimestampProto(ts)
		err = errors.Annotate(err, "encoding time").Err()
	}
	return
}

// Proto converts this git.Commit to its protobuf equivalent.
func (c *Commit) Proto() (ret *git.Commit, err error) {
	ret = &git.Commit{}

	if ret.Id, err = hex.DecodeString(c.Commit); err != nil {
		err = errors.Annotate(err, "decoding id").Err()
		return
	}
	if ret.Tree, err = hex.DecodeString(c.Tree); err != nil {
		err = errors.Annotate(err, "decoding tree").Err()
		return
	}
	if len(c.Parents) > 0 {
		ret.Parents = make([][]byte, len(c.Parents))
		for i, p := range c.Parents {
			if ret.Parents[i], err = hex.DecodeString(p); err != nil {
				err = errors.Annotate(err, "decoding parent %d", i).Err()
				return
			}
		}
	}
	if ret.Author, err = c.Author.Proto(); err != nil {
		err = errors.Annotate(err, "decoding author").Err()
		return
	}
	if ret.Committer, err = c.Committer.Proto(); err != nil {
		err = errors.Annotate(err, "decoding committer").Err()
		return
	}
	ret.Message = c.Message

	// TODO(iannucci): treediff

	return
}

// LogProto takes log data from e.g. Log(), and returns the LUCI standard
// protobuf message form of the log data (suitable for embedding in other proto
// messages).
//
// Errors may occur if the Commits contain bad hashes (i.e. not hexadecimal), or
// undecodable timestamps. This should not occur if the []Commit was produced by
// the Log function in this package.
func LogProto(log []Commit) (ret []*git.Commit, err error) {
	ret = make([]*git.Commit, len(log))
	for i, c := range log {
		if ret[i], err = c.Proto(); err != nil {
			return nil, errors.Annotate(err, "converting commit %d", i).Err()
		}
	}
	return ret, nil
}
