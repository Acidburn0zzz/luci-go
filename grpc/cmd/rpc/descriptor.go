// Copyright 2016 The LUCI Authors. All rights reserved.
// Use of this source code is governed under the Apache License, Version 2.0
// that can be found in the LICENSE file.

package main

import (
	"fmt"
	"strings"

	"github.com/luci/luci-go/common/proto/google/descutil"

	"golang.org/x/net/context"

	"github.com/luci/luci-go/grpc/discovery"
	"github.com/luci/luci-go/grpc/prpc"
	"google.golang.org/genproto/protobuf"
)

type serverDescription struct {
	*discovery.DescribeResponse
}

func loadDescription(c context.Context, client *prpc.Client) (*serverDescription, error) {
	dc := discovery.NewDiscoveryPRPCClient(client)
	res, err := dc.Describe(c, &discovery.Void{})
	if err != nil {
		return nil, fmt.Errorf("could not load server description: %s", err)
	}

	return &serverDescription{res}, nil
}

// resolveInputMessage resolves input message type of a method.
func (d *serverDescription) resolveInputMessage(service, method string) (*descriptor.DescriptorProto, error) {
	_, obj, _ := descutil.Resolve(d.Description, service)
	serviceDesc, ok := obj.(*descriptor.ServiceDescriptorProto)
	if !ok {
		return nil, fmt.Errorf("service %q not found", service)
	}

	mi := descutil.FindMethodForService(serviceDesc, method)
	if mi == -1 {
		return nil, fmt.Errorf("method %q in service %q not found", method, service)
	}

	msgName := serviceDesc.Method[mi].GetInputType()
	msgName = strings.TrimPrefix(msgName, ".")
	return d.resolveMessage(msgName)
}

func (d *serverDescription) resolveMessage(name string) (*descriptor.DescriptorProto, error) {
	_, obj, _ := descutil.Resolve(d.Description, name)
	msg, ok := obj.(*descriptor.DescriptorProto)
	if !ok {
		return nil, fmt.Errorf("message %q not found", name)
	}
	return msg, nil
}
