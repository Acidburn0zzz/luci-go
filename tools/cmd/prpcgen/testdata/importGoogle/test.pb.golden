package test

import prpc "github.com/luci/luci-go/server/prpc"

import proto "github.com/golang/protobuf/proto"
import fmt "fmt"
import math "math"
import google_protobuf "github.com/luci/luci-go/common/proto/google/descriptor"
import google_protobuf1 "github.com/luci/luci-go/common/proto/google"

import (
	context "golang.org/x/net/context"
	grpc "google.golang.org/grpc"
)

var _ = proto.Marshal
var _ = fmt.Errorf
var _ = math.Inf

type Void struct {
}

func (m *Void) Reset()				{ *m = Void{} }
func (m *Void) String() string			{ return proto.CompactTextString(m) }
func (*Void) ProtoMessage()			{}
func (*Void) Descriptor() ([]byte, []int)	{ return fileDescriptor0, []int{0} }

func init() {
	proto.RegisterType((*Void)(nil), "test.Void")
}

var _ context.Context
var _ grpc.ClientConn

type GreeterClient interface {
	Describe(ctx context.Context, in *Void, opts ...grpc.CallOption) (*google_protobuf.FileDescriptorProto, error)
	Timestamp(ctx context.Context, in *Void, opts ...grpc.CallOption) (*google_protobuf1.Timestamp, error)
}

type greeterClient struct {
	cc *grpc.ClientConn
}

func NewGreeterClient(cc *grpc.ClientConn) GreeterClient {
	return &greeterClient{cc}
}

func (c *greeterClient) Describe(ctx context.Context, in *Void, opts ...grpc.CallOption) (*google_protobuf.FileDescriptorProto, error) {
	out := new(google_protobuf.FileDescriptorProto)
	err := grpc.Invoke(ctx, "/test.Greeter/Describe", in, out, c.cc, opts...)
	if err != nil {
		return nil, err
	}
	return out, nil
}

func (c *greeterClient) Timestamp(ctx context.Context, in *Void, opts ...grpc.CallOption) (*google_protobuf1.Timestamp, error) {
	out := new(google_protobuf1.Timestamp)
	err := grpc.Invoke(ctx, "/test.Greeter/Timestamp", in, out, c.cc, opts...)
	if err != nil {
		return nil, err
	}
	return out, nil
}

type GreeterServer interface {
	Describe(context.Context, *Void) (*google_protobuf.FileDescriptorProto, error)
	Timestamp(context.Context, *Void) (*google_protobuf1.Timestamp, error)
}

func RegisterGreeterServer(s prpc.Registrar, srv GreeterServer) {
	s.RegisterService(&_Greeter_serviceDesc, srv)
}

func _Greeter_Describe_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error) (interface{}, error) {
	in := new(Void)
	if err := dec(in); err != nil {
		return nil, err
	}
	out, err := srv.(GreeterServer).Describe(ctx, in)
	if err != nil {
		return nil, err
	}
	return out, nil
}

func _Greeter_Timestamp_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error) (interface{}, error) {
	in := new(Void)
	if err := dec(in); err != nil {
		return nil, err
	}
	out, err := srv.(GreeterServer).Timestamp(ctx, in)
	if err != nil {
		return nil, err
	}
	return out, nil
}

var _Greeter_serviceDesc = grpc.ServiceDesc{
	ServiceName:	"test.Greeter",
	HandlerType:	(*GreeterServer)(nil),
	Methods: []grpc.MethodDesc{
		{
			MethodName:	"Describe",
			Handler:	_Greeter_Describe_Handler,
		},
		{
			MethodName:	"Timestamp",
			Handler:	_Greeter_Timestamp_Handler,
		},
	},
	Streams:	[]grpc.StreamDesc{},
}

var fileDescriptor0 = []byte{

	0x1f, 0x8b, 0x08, 0x00, 0x00, 0x09, 0x6e, 0x88, 0x02, 0xff, 0xe2, 0xe2, 0x2a, 0x49, 0x2d, 0x2e,
	0xd1, 0x2b, 0x28, 0xca, 0x2f, 0xc9, 0x17, 0x62, 0x01, 0xb1, 0xa5, 0x14, 0xd2, 0xf3, 0xf3, 0xd3,
	0x73, 0x52, 0xf5, 0xc1, 0x62, 0x49, 0xa5, 0x69, 0xfa, 0x29, 0xa9, 0xc5, 0xc9, 0x45, 0x99, 0x05,
	0x25, 0xf9, 0x45, 0x10, 0x75, 0x52, 0xf2, 0xe8, 0x2a, 0x4a, 0x32, 0x73, 0x81, 0x3a, 0x13, 0x73,
	0x0b, 0x20, 0x0a, 0x94, 0xd8, 0xb8, 0x58, 0xc2, 0xf2, 0x33, 0x53, 0x8c, 0x1a, 0x18, 0xb9, 0xd8,
	0xdd, 0x8b, 0x52, 0x53, 0x4b, 0x52, 0x8b, 0x84, 0xec, 0xb8, 0x38, 0x5c, 0xc0, 0x06, 0x25, 0xa5,
	0x0a, 0x71, 0xe9, 0x81, 0x6d, 0x05, 0xa9, 0x91, 0x52, 0xd1, 0x83, 0x98, 0xa6, 0x07, 0x33, 0x4d,
	0xcf, 0x2d, 0x33, 0x27, 0xd5, 0x05, 0x6e, 0x67, 0x00, 0xd8, 0x44, 0x06, 0x21, 0x53, 0x2e, 0xce,
	0x10, 0x98, 0x35, 0x28, 0x06, 0x48, 0x61, 0x18, 0x00, 0x57, 0xa7, 0xc4, 0x90, 0xc4, 0x06, 0x16,
	0x35, 0x06, 0x04, 0x00, 0x00, 0xff, 0xff, 0x07, 0x4b, 0x26, 0x50, 0xe8, 0x00, 0x00, 0x00,
}
